lib/rules/no-restricted-properties.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * @fileoverview Rule to disallow certain object properties
+ * @author Will Klein & Eli White
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "disallow certain properties on certain objects",
+            category: "Node.js and CommonJS",
+            recommended: false
+        },
+
+        schema: {
+            type: "array",
+            items: {
+                type: "object",
+                properties: {
+                    object: {
+                        type: "string"
+                    },
+                    property: {
+                        type: "string"
+                    },
+                    message: {
+                        type: "string"
+                    }
+                },
+                additionalProperties: false,
+                required: [
+                    "object",
+                    "property"
+                ]
+            },
+            uniqueItems: true
+        }
+    },
+
+    create(context) {
+        const restrictedCalls = context.options;
+
+        if (restrictedCalls.length === 0) {
+            return {};
+        }
+
+        const restrictedProperties = context.options.reduce(function(restrictions, option) {
+            const objectName = option.object;
+            const propertyName = option.property;
+
+            restrictions[objectName] = restrictions[objectName] || {};
+            restrictions[objectName][propertyName] = {
+                message: option.message
+            };
+
+            return restrictions;
+        }, {});
```

Comment:
I think we should use `Map` objects instead of plain objects.

```js
/*eslint no-restricted-properties: [error, {object: "obj", property: "foo"}]*/

// should not warn.
toString.toString
obj.toString
obj.valueOf
```

```js
/*eslint no-restricted-properties: [error, {object: "obj", property: "__proto__"}]*/

// should not warn.
obj.toString
obj.valueOf
```